File: fyc-api-rest/src/main/java/fr/esgi/fyc_api_rest/mapper/TypeExerciceMapper.java
package fr.esgi.fyc_api_rest.mapper;

import fr.esgi.fyc_api_rest.business.TypeExercice;
import fr.esgi.fyc_api_rest.dto.request.TypeExerciceRequestDTO;
import fr.esgi.fyc_api_rest.dto.response.TypeExerciceResponseDTO;
import fr.esgi.fyc_api_rest.service.CategorieExerciceService;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingConstants;
import org.mapstruct.ReportingPolicy;

@Mapper(
        unmappedTargetPolicy = ReportingPolicy.IGNORE,
        componentModel = MappingConstants.ComponentModel.SPRING,
        uses = {CategorieExerciceService.class, CategorieExerciceMapper.class}
)
public interface TypeExerciceMapper {
    @Mapping(target = "categorie", source = "idCategorie")
    public TypeExercice toEntity(TypeExerciceRequestDTO typeExerciceCreationDTO);

    @Mapping(target = "categorieExerciceResponse", source = "categorie")
    public TypeExerciceResponseDTO toResponseDTO(TypeExercice typeExercice);
}

File: fyc-rest-api-csharp-test/CometeAPI/Mapper/TypeExerciceMapper.cs
using AutoMapper;
using FycRestApi.Business;
using FycRestApi.DTO.Request;
using FycRestApi.DTO.Response;
using FycRestApi.Service.Interface;

namespace FycRestApi.Mapper;

public class TypeExerciceMapper
{
    private readonly IMapper _mapper;
    private readonly ICategorieExerciceService _categorieExerciceService;

    public TypeExerciceMapper(IMapper mapper, ICategorieExerciceService categorieExerciceService)
    {
        this._mapper = mapper;
        this._categorieExerciceService = categorieExerciceService;
    }

    public async Task<TypeExercice> toEntity(TypeExerciceRequestDTO requestDTO)
    {
        TypeExercice typeExercice = _mapper.Map<TypeExercice>(requestDTO);
        typeExercice.categorieExercice = await _categorieExerciceService.findById(requestDTO.idCategorie);
        return typeExercice;
    }

    public TypeExerciceResponseDTO toDTO(TypeExercice typeExercice)
    {
        TypeExerciceResponseDTO typeExerciceResponseDTO = _mapper.Map<TypeExerciceResponseDTO>(typeExercice);
        typeExerciceResponseDTO.categorieExerciceResponse = _mapper.Map<CategorieExerciceResponseDTO>(typeExercice.categorieExercice);
        return typeExerciceResponseDTO;
    }
}